secboot/preinstall_sb.go

```diff
@@ -0,0 +1,137 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//go:build !nosecboot
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"context"
+	"fmt"
+
+	sb_efi "github.com/snapcore/secboot/efi"
+	sb_preinstall "github.com/snapcore/secboot/efi/preinstall"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snapdenv"
+	"github.com/snapcore/snapd/systemd"
+)
+
+var (
+	sbPreinstallNewRunChecksContext = sb_preinstall.NewRunChecksContext
+	sbPreinstallRunChecks           = (*sb_preinstall.RunChecksContext).Run
+)
+
+// PreinstallCheck runs preinstall checks using default check configuration and
+// TCG-compliant PCR profile generation options to evaluate whether the host
+// environment is an EFI system suitable for TPM-based Full Disk Encryption. On
+// success, it returns a list with information on all errors identified by
+// secboot or nil if no errors were found. Any warnings contained in the secboot
+// result are logged. On failure, it returns the error encountered while
+// interpreting the secboot error.
+//
+// To support testing, when test mode is detected and the system is running in a
+// Virtual Machine, the check configuration is modified to permit this without
+// treating it as an error.
+func PreinstallCheck(bootImagePaths []string) ([]PreinstallErrorInfo, error) {
+	// do not customize check configuration
+	checkFlags := sb_preinstall.CheckFlagsDefault
+	if snapdenv.Testing() && systemd.IsVirtualMachine() {
```

Comment:
isn't it always the case that we should permit virtual machine if we are on one?